internal/proxy/proxy.go

```diff
@@ -91,6 +119,77 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusInternalServerError)
 }
 
+func checkEndpoint(url string, kind ProxyKind) error {
+	switch kind {
+	case RPC:
+		return checkRPC(url)
+	case Rest:
+		return checkREST(url)
+	default:
+		return fmt.Errorf("unsupported proxy kind: %v", kind)
+	}
+}
+
+func performGetRequest(url string, timeout time.Duration) ([]byte, error) {
+	client := &http.Client{Timeout: timeout}
+	resp, err := client.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("error making request: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %v", err)
+	}
+
+	return body, nil
+}
+
+func checkRPC(url string) error {
+	body, err := performGetRequest(url+"/status", 2*time.Second)
+	if err != nil {
+		return err
+	}
+
+	var status StatusResponse
+	if err := json.Unmarshal(body, &status); err != nil {
+		return fmt.Errorf("error unmarshaling JSON: %v", err)
+	}
+
+	if status.Result.SyncInfo.CatchingUp {
+		return fmt.Errorf("node is still catching up")
+	}
+
+	if !status.Result.SyncInfo.LatestBlockTime.After(time.Now().UTC().Add(-time.Minute)) {
+		return fmt.Errorf("latest block time is more than 1 minute old")
+	}
+
+	return nil
+}
+
+func checkREST(url string) error {
+	body, err := performGetRequest(url+"/blocks/latest", 2*time.Second)
+	if err != nil {
+		return err
+	}
+
+	var status RestStatusResponse
+	if err := json.Unmarshal(body, &status); err != nil {
+		return fmt.Errorf("error unmarshaling JSON: %v", err)
```

Comment:
turn `error unmarshaling JSON` into something descriptive, it lacks caller context